src/librustc_codegen_ssa/mir/rvalue.rs

```diff
@@ -224,30 +224,7 @@ impl<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>> FunctionCx<'a, 'tcx, Bx> {
                         operand.val
                     }
                     mir::CastKind::Pointer(PointerCast::Unsize) => {
-                        assert!(bx.cx().is_backend_scalar_pair(cast));
-                        match operand.val {
-                            OperandValue::Pair(lldata, llextra) => {
-                                // unsize from a fat pointer -- this is a
-                                // "trait-object-to-supertrait" coercion, for
-                                // example, `&'a fmt::Debug + Send => &'a fmt::Debug`.
-
-                                // HACK(eddyb) have to bitcast pointers
-                                // until LLVM removes pointee types.
-                                let lldata = bx.pointercast(lldata,
-                                    bx.cx().scalar_pair_element_backend_type(cast, 0, true));
-                                OperandValue::Pair(lldata, llextra)
-                            }
-                            OperandValue::Immediate(lldata) => {
-                                // "standard" unsize
-                                let (lldata, llextra) = base::unsize_thin_ptr(&mut bx, lldata,
-                                    operand.layout.ty, cast.ty);
-                                OperandValue::Pair(lldata, llextra)
-                            }
-                            OperandValue::Ref(..) => {
-                                bug!("by-ref operand {:?} in `codegen_rvalue_operand`",
-                                     operand);
-                            }
-                        }
+                        base::coerce_ptr_unsized(&mut bx, operand, cast)
```

Comment:
Hmm, it might be useful to pull this change, without any of the trait upcasting feature, into a separate PR, so we can discuss there the best organization (ideally it would just be moving code around, while making it easy to implement this PR).

Reply:
It's a worthwhile discussion, but please just open an issue about this. Optimisations can come later, as long as we have something working now. I believe it's Niko and my intention to get this merged ASAP now.

Reply:
This is not an optimization, I'm not comfortable landing this sizeable feature PR with several refactors embedded in it.

Reply:
But that's not what I wrote. This is a change to get trait upcasting working. Whether this can be optimised further seems like a separate point... Anyway, Niko seemed happy with this before, so let's see what he says.

Reply:
Well, I definitely want to get this landed, but I also agree with @eddyb that it would be better if we can extract refactorings and land them independently. This PR is definitely very large and it would be great if we could land it in pieces, each one being relatively easy to comprehend.

Reply:
@nikomatsakis Fair enough,  but this one is very much intertwined with functional enhancements to support trait upcasting. Do you have another refactoring in mind here?